metals/src/main/scala/scala/meta/internal/process/SystemProcess.scala

```diff
@@ -1,30 +1,109 @@
 package scala.meta.internal.process
 
-import java.nio.file.Path
+import java.io.IOException
+import java.io.InputStream
+import java.util.concurrent.TimeUnit
 
-import scala.sys.process._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.jdk.CollectionConverters._
+import scala.sys.process.BasicIO
+import scala.util.control.NonFatal
 
-import scala.meta.internal.metals.Time
-import scala.meta.internal.metals.Timer
-import scala.meta.pc.CancelToken
+import scala.meta.internal.ansi.AnsiFilter
+import scala.meta.io.AbsolutePath
+
+trait SystemProcess {
+  def complete: Future[Int]
+  def cancel: Unit
+}
 
 object SystemProcess {
+
   def run(
-      shortName: String,
-      args: List[String],
-      reproduceArgs: List[String],
-      cwd: Path,
-      token: CancelToken
-  ): Unit = {
-    val exportTimer = new Timer(Time.system)
-    scribe.info(args.mkString("process: ", " ", ""))
-    val exit = Process(args, cwd = Some(cwd.toFile())).!
-    if (exit != 0) {
-      val message = s"$shortName command failed with exit code $exit, " +
-        s"to reproduce run the command below:\n\t${reproduceArgs.mkString(" ")}"
-      throw new MessageOnlyException(message)
-    } else {
-      scribe.info(s"time: ran '$shortName' in $exportTimer")
+      cmd: List[String],
+      cwd: AbsolutePath,
+      redirectErrorOutput: Boolean,
+      env: Map[String, String]
+  ): SystemProcess = {
+
+    try {
+      val builder = new ProcessBuilder(cmd.asJava)
+      builder.directory(cwd.toNIO.toFile)
+      val envMap = builder.environment()
+      envMap.putAll(env.asJava)
+
+      builder.redirectErrorStream(redirectErrorOutput)
+      val ps = builder.start()
+      wrapProcess(ps, redirectErrorOutput)
+    } catch {
+      case NonFatal(e) =>
+        scribe.error(s"Running process '${cmd.mkString(" ")}' failed", e)
+        Failed
     }
+
   }
+
+  def wrapProcess(
+      ps: Process,
+      redirectErrorOutput: Boolean
+  ): SystemProcess = {
+    def readOutput(stream: InputStream, f: String => Unit): Thread = {
+      val filter = AnsiFilter()
+      val thread = new Thread {
+        override def run(): Unit = {
+          // use scala.sys.process implementation
+          try {
+            BasicIO.processFully(line => f(filter(line)))(
+              stream
+            )
+          } catch {
+            case _: IOException => // that's ok, happens on cancel
+            case NonFatal(e) =>
+              scribe.error("Unexcepted error in reading out", e)
+          }
+        }
+      }
+      thread.setDaemon(true)
```

Comment:
Do we need it to be a deamon?

Reply:
I think yes. Just to avoid a possibility of having dangling thread that keeps JVM waiting its completion on shutdown.